controllers/onload_controller.go

```diff
@@ -78,34 +90,515 @@ func (r *OnloadReconciler) Reconcile(
+	if err != nil {
+		log.Error(err, "Failed to patch device plugin DaemonSet",
+			"device plugin", devicePlugin)
+		return nil, err
+	}
+
+	log.Info("Patched device plugin Daemonset", "device plugin", devicePlugin)
+	return &ctrl.Result{Requeue: true}, nil
+}
+
+func (r *OnloadReconciler) handleModuleUpdate(ctx context.Context, onload *onloadv1alpha1.Onload) (*ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	module := &kmm.Module{}
+	err := r.Get(ctx, types.NamespacedName{Name: onload.Name + onloadModuleNameSuffix, Namespace: onload.Namespace}, module)
+	if err != nil {
+		log.Error(err, "Failed to get onload module", "Onload", onload)
+		return nil, err
+	}
+
+	if module.Spec.ModuleLoader.Container.Version == onload.Spec.Onload.Version {
+		// Nothing to be done, so return
+		return nil, nil
+	}
+
+	oldModule := module.DeepCopy()
+
+	module.Spec.ModuleLoader.Container.Version = onload.Spec.Onload.Version
+
+	for i := range module.Spec.ModuleLoader.Container.KernelMappings {
+		moduleKmap := &module.Spec.ModuleLoader.Container.KernelMappings[i]
+		for _, onloadKmap := range onload.Spec.Onload.KernelMappings {
+			if onloadKmap.Regexp == moduleKmap.Regexp {
+				moduleKmap.ContainerImage = onloadKmap.KernelModuleImage
+				break
+			}
+		}
+	}
+
+	err = r.Patch(ctx, module, client.MergeFrom(oldModule))
+	if err != nil {
+		log.Error(err, "Failed to patch module", "Module", module)
+		return nil, err
+	}
+
+	log.Info("Updated module definition for upgrade", "Module", module)
+	return &ctrl.Result{Requeue: true}, nil
+}
+
+const defaultRequeueTime = 5 * time.Second
+
+func (r *OnloadReconciler) handleUpdate(ctx context.Context, onload *onloadv1alpha1.Onload) (*ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	res, err := r.handleDevicePluginUpdate(ctx, onload)
+	if err != nil || res != nil {
+		return res, err
+	}
+
+	res, err = r.handleModuleUpdate(ctx, onload)
+	if err != nil || res != nil {
+		return res, err
+	}
+
+	badNodes, err := r.getBadNodes(ctx, onload)
+	if err != nil {
+		return nil, err
+	}
+	if len(badNodes) == 0 {
+		// Nothing to be done, so return
+		return nil, nil
+	}
+
+	// We just want to upgrade a single node at a time.
+	// Since I don't know if there are any guarantees about how they are ordered
+	// when returned from List(), we shall upgrade nodes by their name
+	// alphabetically.
+	// Golang 1.21 introduces the "slices" package which would allow us to find
+	// the min element of a slice in a single line, however at the moment the
+	// operator is written in Go 1.19 so this package isn't available.
+	// In it's absence I am just using a local funcion.
+
+	node := func(nodes []corev1.Node) corev1.Node {
+		min := nodes[0]
+		for _, node := range nodes {
+			if node.Name < min.Name {
+				min = node
+			}
+		}
+		return min
+	}(badNodes)
+
+	log.Info("Updating onload version on node "+node.Name, "onload", onload)
+
+	// Remove the onload label from the node
```

Comment:
Also, I think it's worth adding a log entry indicating progress, e.g. how many nodes are on the list to upgrade, what do you think?

Reply:
I did have some logging (e.g. "There are N Nodes with a bad onload version"), but since it would be logged every time there was a version-mismatch it would end up spamming the log.
I'll look into updating the status to reflect the upgrade process